### ISSUE TITLE
OLMoE-7A1B QKNorm Architecture Incompatibility with Megatron-Core

### ISSUE THREAD

**Issue by JungHoyoun** [NVIDIA-NeMo/Megatron-Bridge#1532]
## Describe the bug

OLMoE-7A1B (allenai/OLMoE-1B-7B-0125) models cannot be properly supported in Megatron-Bridge due to a fundamental architectural difference in how QKNorm is applied compared to other models like Qwen3.

**Key Issue**: OLMoE applies QKNorm to the hidden dimension directly, whereas models like Qwen3 apply QKNorm per attention head. Megatron-Core's attention implementation only supports QKNorm at the head level, not at the hidden dimension level.

## Steps/Code to reproduce bug

1. Convert OLMoE-7A1B model using Megatron-Bridge:
```
python examples/conversion/convert_checkpoints.py import \
  --hf-model allenai/OLMoE-1B-7B-0125 \
  --megatron-path ./checkpoints/olmoe-1b-7b-0125 \
```

2. Run the compare code:
```
torchrun --nproc_per_node=1 examples/conversion/compare_hf_and_megatron/compare.py \
    --hf_model_path "allenai/OLMoE-1B-7B-0125" \
    --megatron_model_path "./checkpoints/olmoe-1b-7b-0125" \
    --prompt "Hello world" 
```
TBD

## Expected behavior
it would be better to explicitly exclude OLMoE-7A1B from Megatron-Bridge support rather than attempting to implement workarounds. The required changes to Megatron-Core's attention layer would be too invasive.

### Architecture Comparison

**Qwen3/Qwen-like models** (supported):
- QKNorm is applied per attention head
- Reference: [Qwen3 attention implementation](https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen3/modeling_qwen3.py#L246-L251)
- Megatron-Core supports this via `q_layernorm` and `k_layernorm` applied after Q/K projection

**OLMoE-7A1B** (not supported):
- QKNorm is applied to the full hidden dimension before Q/K/V projection
- Reference: [Olmoe attention implementation](https://github.com/huggingface/transformers/blob/main/src/transformers/models/olmoe/modeling_olmoe.py#L258-L260)
- Megatron-Core's attention implementation does not support this pattern
- Reference: [Megatron attention implementation](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/attention.py#L1149-L1153)

**Comment by suiyoubi** [comment]
Hi @JungHoyoun, thanks for creating this issue. We currently have the custom attention logic for MLMoE implemented in Megatron-Bridge:
https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/c3e43e6aa8d6a3bb52a3e2d2a34ef484eb90bd7c/src/megatron/bridge/models/olmoe/olmoe_provider.py#L98

This is the typical approach we take when supporting models that differ slightly from the standard Megatron-Core implementation. I’m curious whether you’ve run into any problems with this setup — for example, mismatched logits, unexpected loss behavior, or other inconsistencies?

I’m trying to understand the statement: “_The required changes to Megatron-Core's attention layer would be too invasive._” Could you provide more context on what specifically felt too invasive?

**Comment by JungHoyoun** [comment]
Hi, thanks for the detailed explanation — and sorry for the confusion earlier.

In my environment, the model was unexpectedly going through the Megatron-Core attention path rather than the custom OLMoE attention implementation. Because of that, I mistakenly assumed that handling OLMoE’s hidden-dim QKNorm would require invasive changes to Megatron-Core itself.

Given your clarification and the presence of the custom OLMoESelfAttention in Megatron-Bridge, it seems my observation was due to a misconfiguration or an environment-specific issue on my side, not a limitation of the framework. Sorry about the misleading statement.

I’ll re-check my setup. Thanks again for the clarification!

**Comment by pruprakash** [comment]
## QA RCCA Analysis

**Classification:** USER ERROR / MISCONFIGURATION

**Summary:** User reported that OLMoE-7A1B models could not be supported in Megatron-Bridge due to QKNorm architectural differences with Megatron-Core. However, investigation revealed that Megatron-Bridge already has custom attention logic for OLMoE that handles this difference.

**Root Cause:** User misconfiguration caused the model to use the Megatron-Core attention path instead of the custom `OLMoESelfAttention` implementation in Megatron-Bridge. The user confirmed this was an environment-specific issue on their side.

**Resolution:** User acknowledged the misunderstanding after maintainer pointed to the existing custom OLMoE attention implementation at `src/megatron/bridge/models/olmoe/olmoe_provider.py`. No code changes were required.

**Test Coverage Verdict:** N/A

**Rationale:** This is not a code bug. The feature (OLMoE support with custom QKNorm handling) was already implemented and working. The issue was due to user misconfiguration causing incorrect code path execution. User configuration issues are outside the scope of unit/functional test coverage.
